test/test_common/simulated_time_system.h

```diff
@@ -64,10 +65,27 @@ class SimulatedTimeSystemHelper : public TestTimeSystem {
   class SimulatedScheduler;
   class Alarm;
   friend class Alarm; // Needed to reference mutex for thread annotations.
-  struct CompareAlarms {
-    bool operator()(const Alarm* a, const Alarm* b) const;
+  struct AlarmRegistration {
+    AlarmRegistration(MonotonicTime time, uint64_t randomness, Alarm* alarm)
+        : time_(time), randomness_(randomness), alarm_(alarm) {}
+
+    MonotonicTime time_;
+    // Random tie-breaker for alarms scheduled for the same monotonic time used to mimic
+    // non-deterministic execution of real alarms scheduled for the same wall time.
+    uint64_t randomness_;
+    Alarm* alarm_;
+
+    friend bool operator<(const AlarmRegistration& lhs, const AlarmRegistration& rhs) {
+      if (lhs.time_ != rhs.time_) {
+        return lhs.time_ < rhs.time_;
+      }
+      if (lhs.randomness_ != rhs.randomness_) {
+        return lhs.randomness_ < rhs.randomness_;
+      }
+      return lhs.alarm_ < rhs.alarm_;
```

Comment:
you are intentionally comparing pointers here?

Reply:
yes.  This is a final tie breaker that is almost impossible to hit.

Reply:
I see...it would be hit only if you get two of the same random number. Worth a comment?

Reply:
Also, while this is super-rare, it introduces non-determinism outside the scope of the `--seed` flag. You could also retain `index_` but just have `randomness_` supersede.

that would also allow users to specify `--seed 0` and disable the non-determinism, in case someone wants that while debugging. For seed==0 we'd just populate `.randomness=0` and fall-through to index.

Reply:
--seed=0 already means "use a random seed".  The odds of randomness_ being equal are close to 1 in 2**64 since the number of alarms used in a test is relatively small.

Reply:
Yeah my argument was pedantic; this is OK but IMO worth a comment.

Reply:
Added comment and ASSERT.